ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/TransactionLogsIndexer.java

```diff
@@ -117,16 +128,68 @@ private long fillCacheFile(
       if (maybeHeader.isEmpty()) {
         break;
       }
-      final byte[] logs = maybeHeader.get().getLogsBloom().toArray();
-      checkNotNull(logs);
-      checkState(logs.length == 256, "BloomBits are not the correct length");
-      fos.write(logs);
+      fillCacheFileWithBlock(maybeHeader.get(), fos);
       indexingStatus.currentBlock = blockNum;
       blockNum++;
     }
     return blockNum - startBlock;
   }
 
+  public void cacheLogsBloomForBlockHeader(final BlockHeader blockHeader) {
+    try {
+      final long blockNumber = blockHeader.getNumber();
+      LOG.info("Caching logs bloom for block {}.", "0x" + Long.toHexString(blockNumber));
+      ensurePreviousSegmentsArePresent(blockNumber);
+      final File cacheFile = calculateCacheFileName(blockNumber, cacheDir);
+      if (!cacheFile.exists()) {
+        Files.createFile(cacheFile.toPath());
+      }
+      try (RandomAccessFile writer = new RandomAccessFile(cacheFile, "rw")) {
+        final long offset = (blockNumber / BLOCKS_PER_BLOOM_CACHE) * BLOOM_BITS_LENGTH;
+        writer.seek(offset);
+        writer.write(ensureBloomBitsAreCorrectLength(blockHeader.getLogsBloom().toArray()));
+      }
+    } catch (IOException e) {
+      LOG.error("Unhandled indexing exception.", e);
+    }
+  }
+
+  private void ensurePreviousSegmentsArePresent(final long blockNumber) {
+    scheduler.scheduleFutureTask(
```

Comment:
We should check indexing status before submitting the work.

Reply:
Done.